contracts/RelayedDocumentStore.sol

```diff
@@ -0,0 +1,112 @@
+  *         OR
+  *         - const documentMerkleRoots = ["0x3a267813bea8120f55a7b9ca814c34dd89f237502544d7c75dfd709a659f6330", "...", ...];
+  *           const message = web3.utils.soliditySha3(...documentMerkleRoots); // hash of many documentMerkleRoots for a batch operation
+  *       2. Generate the nonce... transaction count... random number etc.
+  *       3. Generate the signature.  This signature is generated by signing the hash of: the nonce, the message as well as the relay contract addres.
+  *          The relay contract address is used to also elimiate the risk of replay attacks across other relay contracts.
+  *          The resulting signature should be in hex format following the Ethereum signature encoding.
+  *          For example:
+  *           - const message = "0x3a267813bea8120f55a7b9ca814c34dd89f237502544d7c75dfd709a659f6330";
+  *             const nonce = 1;
+  *             const address = contract.address; // 0x8CdaF0CD259887258Bc13a92C0a6dA92698644C0
+  *             const messageHash = web3.utils.soliditySha3(message, nonce, address); // 0x2bd7dce779d8d21e3266e847f21ce8160a53e6a78a65a5045f72b18081f7fc69
+  *             const signature = await web3.eth.sign(messageHash, signer); // 0xf5731da089e7d9532aad446f9f92cabf3ce39c5cde8ba541cb26ed17b69ab8870a411a66ad8316d7b7478e1c1be05ca840ecfc8cc7468beb9c2832be4dc8fb1e1b
+ */
+contract RelayedDocumentStore is DocumentStore {
+  using ECDSA for bytes32;
+
+  address private trustedSigner; // Account that is signing *messages*
+
+  function initialize(
+    string memory _name,
+    address _trustedSigner,
+    address _relayer
+  )
+    public
+    initializer
+  {
+    super.initialize(_name, _relayer);
+    require(_trustedSigner != address(0), "RelayedDocumentStore: trusted signer is the zero address");
+    trustedSigner = _trustedSigner;
+  }
+
+  function bulkIssueRelayed(
+    bytes32[] memory documents,
+    bytes32 documentsHash,
+    uint256 nonce,
+    bytes memory signature
+  )
+    public
+    onlyTrustedSigner(documentsHash, nonce, signature)
+  {
+    bulkIssue(documents);
+  }
+
+  function bulkRevokeRelayed(
+    bytes32[] memory documents,
+    bytes32 documentsHash,
+    uint256 nonce,
+    bytes memory signature
+  )
+    public
+    onlyTrustedSigner(documentsHash, nonce, signature)
+  {
+    bulkRevoke(documents);
+  }
+
+  function issueRelayed(
+    bytes32 document,
+    uint256 nonce,
+    bytes memory signature
+  )
+    public
+    onlyTrustedSigner(document, nonce, signature)
+  {
+    issue(document);
+  }
+
+  function revokeRelayed(
+    bytes32 document,
+    uint256 nonce,
+    bytes memory signature
+  )
+    public
+    onlyTrustedSigner(document, nonce, signature)
+  {
+    revoke(document);
+  }
+
+  /**
+   * @dev Throws if the message is not signed by the trusted signer.
+   */
+  modifier onlyTrustedSigner(bytes32 message, uint256 nonce, bytes memory signature) {
```

Comment:
recommend adding an event log here to log that <tx> was relayed by <relayer> on behalf of <signer>

Reply:
The following has been added:
```
event RelayedMessage(address indexed relayer, address indexed trustedSigner, bytes4 functionIdentifier);
```

Invoked within the modifier as such:
```
emit RelayedMessage(msg.sender, recoveredAddress, msg.sig);
```